### PR TITLE
Service node checkpointing deregistration

### PR DESCRIPTION
**Pull request by Doy-lee**
Allow deregisters by counting the number of expected votes to have been received and the number of votes we actually receive. It's a bit of an awkward API in that since we want to store this information into the service node list from quorum cop you need to make an accessor from quorum_cop -> core -> service_node_list to increment the expected amount of votes and counting the received votes.

Before any deregistration occurs, the service node must be online for atleast 2 hours so it collects the necessary voting information and doesn't deregister too prematurely. A service node is required to vote in atleast 8 checkpoints before we check how many votes we've actually received from them.

They are allowed to miss up to 4 votes before we consider them for deregistration.

### CODE REVIEW

File: src/cryptonote_core/service_node_list.cpp [hunk 1]
```diff
@@ -1537,19 +1557,206 @@ namespace service_nodes
...
+
+    crypto::cn_fast_hash(buf, sizeof(buf), result);
+
+    return result;
+  }
+
+  cryptonote::NOTIFY_UPTIME_PROOF::request service_node_list::generate_uptime_proof(crypto::public_key const &pubkey,
+                                                                                    crypto::secret_key const &key,
+                                                                                    uint32_t public_ip,
+                                                                                    uint32_t storage_port) const
```
Comment (msgmaxim) [inline src/cryptonote_core/service_node_list.cpp]: should be `uint16_t` for port

File: src/cryptonote_core/service_node_list.cpp [hunk 2]
```diff
@@ -1537,19 +1557,206 @@ namespace service_nodes
...
+    // NOTE: Validate proof version, timestamp range,
+    {
+      if ((proof.timestamp < now - UPTIME_PROOF_BUFFER_IN_SECONDS) || (proof.timestamp > now + UPTIME_PROOF_BUFFER_IN_SECONDS))
+      {
+        LOG_PRINT_L2("Rejecting uptime proof from " << proof.pubkey << ": timestamp is too far from now");
+        return false;
+      }
+
+      // NOTE: Only care about major version for now
+      // FIXME(Jason): remove this `false` before release!
```
Comment (msgmaxim) [inline src/cryptonote_core/service_node_list.cpp]: remove it in this PR?

File: src/cryptonote_core/service_node_list.cpp (hunk 1 above)
Comment (Doy-lee) [reply src/cryptonote_core/service_node_list.cpp]: Fixed

File: src/cryptonote_core/service_node_list.cpp (hunk 2 above)
Comment (Doy-lee) [reply src/cryptonote_core/service_node_list.cpp]: Yes removed

**Pull request merged**